src/Client/GetHedgedConnections.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <Common/TimerDescriptor.h>
+#include <Client/ConnectionPoolWithFailover.h>
+#include <Core/Settings.h>
+#include <Common/Epoll.h>
+#include <unordered_map>
+#include <memory>
+
+namespace DB
+{
+
+using TimerDescriptorPtr = std::shared_ptr<TimerDescriptor>;
+
+/// Class for establishing hedged connections with replicas.
+/// It works with multiple replicas simultaneously without blocking by using epoll.
+class GetHedgedConnections
+{
+public:
+    using ShuffledPool = ConnectionPoolWithFailover::Base::ShuffledPool;
+
+    enum State
+    {
+        EMPTY = 0,
+        READY = 1,
+        NOT_READY = 2,
+        CANNOT_CHOOSE = 3,
+    };
+
+    struct ReplicaState
+    {
+        Connection * connection = nullptr;
+        State state = State::EMPTY;
+        int index = -1;
+        int fd = -1;
+        size_t parallel_replica_offset = 0;
+        std::unordered_map<int, std::shared_ptr<TimerDescriptor>> active_timeouts;
+
+        void reset()
+        {
+            connection = nullptr;
+            state = State::EMPTY;
+            index = -1;
+            fd = -1;
+            parallel_replica_offset = 0;
+            active_timeouts.clear();
+        }
+
+        bool isReady() const { return state == State::READY; }
+        bool isNotReady() const { return state == State::NOT_READY; }
+        bool isEmpty() const { return state == State::EMPTY; }
+        bool isCannotChoose() const { return state == State::CANNOT_CHOOSE; }
```

Comment:
It is not clear why those methods are needed.
I think it is more common to use `switch/case` for `enum` (`enum class` is better)